File: idea/tests/org/jetbrains/kotlin/idea/refactoring/inline/AbstractInlineTest.java
/*
 * Copyright 2010-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.idea.refactoring.inline;

import com.intellij.codeInsight.TargetElementUtil;
import com.intellij.lang.refactoring.InlineActionHandler;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.extensions.Extensions;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.psi.PsiElement;
import com.intellij.refactoring.util.CommonRefactoringUtil;
import com.intellij.testFramework.LightProjectDescriptor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.kotlin.idea.test.JetLightCodeInsightFixtureTestCase;
import org.jetbrains.kotlin.idea.test.JetWithJdkAndRuntimeLightProjectDescriptor;
import org.jetbrains.kotlin.test.InTextDirectivesUtils;

import java.io.File;
import java.io.IOException;
import java.util.List;

import static com.intellij.codeInsight.TargetElementUtilBase.ELEMENT_NAME_ACCEPTED;
import static com.intellij.codeInsight.TargetElementUtilBase.REFERENCED_ELEMENT_ACCEPTED;

public abstract class AbstractInlineTest extends JetLightCodeInsightFixtureTestCase {
    @Nullable
    private static InlineActionHandler findSuitableHandler(@NotNull PsiElement psiElement) {
        InlineActionHandler[] inlineActionHandlers = Extensions.getExtensions(InlineActionHandler.EP_NAME);
        for (InlineActionHandler handler : inlineActionHandlers) {
            if (handler.canInlineElement(psiElement)) {
                return handler;
            }
        }
        return null;
    }

    protected void doTest(@NotNull String path) throws IOException {
        File afterFile = new File(path + ".after");

        myFixture.configureByFile(path);

        boolean afterFileExists = afterFile.exists();

        final PsiElement targetElement =
                TargetElementUtil.findTargetElement(myFixture.getEditor(), ELEMENT_NAME_ACCEPTED | REFERENCED_ELEMENT_ACCEPTED);

        assertNotNull(targetElement);

        final InlineActionHandler handler = findSuitableHandler(targetElement);

        List<String> expectedErrors = InTextDirectivesUtils.findLinesWithPrefixesRemoved(myFixture.getFile().getText(), "// ERROR: ");
        if (handler != null) {
            try {
                ApplicationManager.getApplication().runWriteAction(new Runnable() {
                    @Override
                    public void run() {
                        handler.inlineElement(myFixture.getProject(), myFixture.getEditor(), targetElement);
                    }
                });

                assertTrue(afterFileExists);
                assertEmpty(expectedErrors);
                myFixture.checkResult(FileUtil.loadFile(afterFile, true));
            } catch (CommonRefactoringUtil.RefactoringErrorHintException e) {
                assertFalse(afterFileExists);
                assertEquals(1, expectedErrors.size());
                assertEquals(expectedErrors.get(0).replace("\\n", "\n"), e.getMessage());
            }
        }
        else {
            assertFalse(afterFileExists);
        }
    }

    @NotNull
    @Override
    protected LightProjectDescriptor getProjectDescriptor() {
        return JetWithJdkAndRuntimeLightProjectDescriptor.INSTANCE;
    }
}

File: idea/tests/org/jetbrains/kotlin/idea/refactoring/inline/AbstractInlineTest.kt
/*
 * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */

package org.jetbrains.kotlin.idea.refactoring.inline

import com.intellij.codeInsight.TargetElementUtil
import com.intellij.codeInsight.TargetElementUtil.ELEMENT_NAME_ACCEPTED
import com.intellij.codeInsight.TargetElementUtil.REFERENCED_ELEMENT_ACCEPTED
import com.intellij.lang.refactoring.InlineActionHandler
import com.intellij.openapi.extensions.Extensions
import com.intellij.openapi.util.io.FileUtil
import com.intellij.refactoring.BaseRefactoringProcessor
import com.intellij.refactoring.util.CommonRefactoringUtil
import com.intellij.testFramework.UsefulTestCase
import com.intellij.testFramework.fixtures.JavaCodeInsightTestFixture
import junit.framework.TestCase
import org.jetbrains.kotlin.idea.test.KotlinLightCodeInsightFixtureTestCase
import org.jetbrains.kotlin.idea.test.KotlinWithJdkAndRuntimeLightProjectDescriptor
import org.jetbrains.kotlin.idea.test.configureCompilerOptions
import org.jetbrains.kotlin.idea.test.rollbackCompilerOptions
import org.jetbrains.kotlin.idea.util.application.runWriteAction
import org.jetbrains.kotlin.test.InTextDirectivesUtils
import org.jetbrains.kotlin.test.KotlinTestUtils
import java.io.File

abstract class AbstractInlineTest : KotlinLightCodeInsightFixtureTestCase() {
    val fixture: JavaCodeInsightTestFixture
        get() = myFixture

    protected fun doTest(unused: String) {
        val mainFile = testDataFile()
        val afterFile = File(testDataPath, "${fileName()}.after")

        val mainFileName = mainFile.name
        val mainFileBaseName = FileUtil.getNameWithoutExtension(mainFileName)
        val extraFiles = mainFile.parentFile.listFiles { _, name ->
            name != mainFileName && name.startsWith("$mainFileBaseName.") && (name.endsWith(".kt") || name.endsWith(".java"))
        }
        val extraFilesToPsi = extraFiles.associateBy { fixture.configureByFile(it.name) }
        val file = myFixture.configureByFile(fileName())

        val configured = configureCompilerOptions(file.text, project, module)

        try {
            val afterFileExists = afterFile.exists()

            val targetElement =
                TargetElementUtil.findTargetElement(myFixture.editor, ELEMENT_NAME_ACCEPTED or REFERENCED_ELEMENT_ACCEPTED)!!

            @Suppress("DEPRECATION")
            val handler = Extensions.getExtensions(InlineActionHandler.EP_NAME).firstOrNull { it.canInlineElement(targetElement) }

            val expectedErrors = InTextDirectivesUtils.findLinesWithPrefixesRemoved(myFixture.file.text, "// ERROR: ")
            if (handler != null) {
                try {
                    runWriteAction { handler.inlineElement(myFixture.project, myFixture.editor, targetElement) }

                    UsefulTestCase.assertEmpty(expectedErrors)
                    KotlinTestUtils.assertEqualsToFile(afterFile, file.text)
                    for ((extraPsiFile, extraFile) in extraFilesToPsi) {
                        KotlinTestUtils.assertEqualsToFile(File("${extraFile.path}.after"), extraPsiFile.text)
                    }
                } catch (e: CommonRefactoringUtil.RefactoringErrorHintException) {
                    TestCase.assertFalse("Refactoring not available: ${e.message}", afterFileExists)
                    TestCase.assertEquals("Expected errors", 1, expectedErrors.size)
                    TestCase.assertEquals("Error message", expectedErrors[0].replace("\\n", "\n"), e.message)
                } catch (e: BaseRefactoringProcessor.ConflictsInTestsException) {
                    TestCase.assertFalse("Conflicts: ${e.message}", afterFileExists)
                    TestCase.assertEquals("Expected errors", 1, expectedErrors.size)
                    TestCase.assertEquals("Error message", expectedErrors[0].replace("\\n", "\n"), e.message)
                }
            } else {
                TestCase.assertFalse("No refactoring handler available", afterFileExists)
            }
        } finally {
            if (configured) {
                rollbackCompilerOptions(project, module)
            }
        }
    }

    override fun getProjectDescriptor() = KotlinWithJdkAndRuntimeLightProjectDescriptor.INSTANCE
}
